File: 백준/Gold/1068. 트리/트리.py
N = int(input())
nodes = list(map(int, input().split()))
del_node = int(input())

def solution(N, nodes, del_node):
    global cnt
    cnt = 0
    adj = [[] for _ in range(N)]
    for idx, num in enumerate(nodes):
        if num == -1 or idx == del_node:
            continue
        adj[num].append(idx) # idx가 노드번호
        
    # 삭제노드 = 방문 물가 > visited 기록
    visited = [False for _ in range(N)]
    visited[del_node] = True
    
    leaf_nodes = []
    def dfs(cur, lv):
        global cnt
        # cnt = 0
        visited[cur] = True # 방문체크
        
        # 리프노드 체크
        if len(adj[cur]) == 0:
            cnt += 1
            leaf_nodes.append(cur)
        # print(f"방문 :{cur}, visited: {visited}, lv :{lv}, cnt: {cnt}")
        for nxt in adj[cur]:
            if not visited[nxt]:
                dfs(nxt, lv + 1)
                # cnt += 1
        # # cnt 똑같이 0이라는건 방문한게 없다는 것 > 리프노드
        # if cnt == 0: 
        #     leaf_nodes.append(cur)

    root = nodes.index(-1)
    
    # print(adj, root)
    if root == del_node:
        return 0
    
    dfs(root, 0)
    
    # print(leaf_nodes)
    
    # return len(leaf_nodes)
    return cnt
    
print(solution(N, nodes, del_node))

File: 백준/Gold/1068. 트리/트리.java
import java.io.*;
import java.util.*;

class Solution {
    int n, delNode;
    int[] nodes;
    int[] visited;
    // 인접행렬
    List<List<Integer>> adj = new ArrayList<>();

    public int solution(int n, int[] nodes, int delNode) {
        this.n = n;
        this.delNode = delNode;
        this.nodes = nodes;
        this.visited = new int[n];

        // 인접행렬 초기화
        for (int i = 0; i < n; i++) {
            adj.add(new ArrayList<>());
        }
        // 인접행렬 생성
        int root = 0;
        for (int i = 0; i < n; i++) {
            if (nodes[i] != -1) {
                adj.get(nodes[i]).add(i);
            } else {
                root = i;
            }
        }

        // 루트 노드를 삭제하는 경우
        if (root == delNode) {
            return 0;
        }


        int cnt = dfs(root, -1);

        return cnt;
    }

    int dfs(int cur, int par) {
        int ret = 0;

        if (visited[cur] == 1) {
            return 0;
        }
        visited[cur] = 1;
        boolean hasChild = false;
        for (int nxt : adj.get(cur)) {
           // 삭제할 아이 만나면 탐색 금지! 
            if (nxt == delNode) {
                continue;
            }
            if (nxt != par) {
                ret += dfs(nxt, cur);
                hasChild = true;
            }
        }
        if (!hasChild) {
            ret++;
        }
        return ret;
    }
}

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());
        int[] nodes = new int[n];

        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            nodes[i] = Integer.parseInt(st.nextToken());
        }

        int delNode = Integer.parseInt(br.readLine());

        Solution sol = new Solution();
        System.out.println(sol.solution(n, nodes, delNode));
    }
}
